### DISCUSSION TITLE
Multi scheduler caused random worker pods crash

### DISCUSSION THREAD

**Discussion by smartnewsdingli** [apache/airflow#20269] [General]
### Apache Airflow version

2.2.2 (latest released)

### What happened

We have multi schedulers setup with kubernetesExecutor. We find of all 14000 tasks run daily, there are 30+ worker pods crashed because of `airflow.exceptions.AirflowException: Hostname of job runner does not match`.
Then we use one scheduler, this issue gone.

### Operating System

ubuntu

### Deployment

Other

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Without logs and reproduction path there is not much anyone can do. 

This looks like a temporary DNS problem with your Kubernetes. But you need to look deeper in your K8S logs to find out.

**Comment by cloventt** [comment]
We see the same issue.
The worker pod prints this log as soon as it starts:
```
Running <TaskInstance: hourly_reporting_bid_requests.ssp-day_hourly-bid-requests scheduled__2022-07-04T13:00:00+00:00 [queued]> on host hourlyreportingbidrequestssspdayhourlybidrequests.05d7664cfd914
Traceback (most recent call last):
  File "/home/airflow/.local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/__main__.py", line 48, in main
    args.func(args)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/cli.py", line 92, in wrapper
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/commands/task_command.py", line 298, in task_run
    _run_task_by_selected_method(args, dag, ti)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/commands/task_command.py", line 105, in _run_task_by_selected_method
    _run_task_by_local_task_job(args, ti)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/commands/task_command.py", line 163, in _run_task_by_local_task_job
    run_job.run()
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/jobs/base_job.py", line 245, in run
    self._execute()
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/jobs/local_task_job.py", line 134, in _execute
    self.heartbeat()
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/jobs/base_job.py", line 226, in heartbeat
    self.heartbeat_callback(session=session)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/session.py", line 67, in wrapper
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/jobs/local_task_job.py", line 196, in heartbeat_callback
    raise AirflowException("Hostname of job runner does not match")
airflow.exceptions.AirflowException: Hostname of job runner does not match
```
We have two schedulers, both running Airflow 2.2.5.

Our worker pod config looks like this:
```yaml
    ---
    apiVersion: v1
    kind: Pod
    metadata:
      name: dummy-name
      namespace: airflow2
    spec:
      serviceAccountName: airflow2
      terminationGracePeriodSeconds: 60
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                    - key: kubernetes_executor
                      operator: In
                      values:
                        - True
                topologyKey: kubernetes.io/hostname
      initContainers:
        - name: airflow2-worker-git-sync-init
          image: k8s.gcr.io/git-sync/git-sync:v3.3.5
          env:
            - name: GIT_SYNC_REPO
              value: <repo url>
            - name: GIT_SYNC_BRANCH
              value: master
            - name: GIT_SYNC_DEST
              value: repo
            - name: GIT_SYNC_WAIT
              value: '1'
            - name: GIT_SYNC_ONE_TIME
              value: 'true'
            - name: GIT_SYNC_USERNAME
              valueFrom:
                secretKeyRef:
                  key: GIT_SYNC_USERNAME
                  name: airflow2
            - name: GIT_SYNC_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: GIT_SYNC_PASSWORD
                  name: airflow2
          resources:
            requests:
              cpu: 10m
              memory: 32Mi
          volumeMounts:
            - mountPath: /tmp/git
              name: dags
      containers:
        - name: airflow2-worker
          image: <image>
          envFrom:
            - secretRef:
                name: airflow2
          volumeMounts:
            - mountPath: /opt/airflow/dags
              name: dags
              readOnly: true
            - mountPath: /opt/airflow/airflow.cfg
              subPath: airflow.cfg
              name: config
              readOnly: true
      restartPolicy: Never
      securityContext:
        runAsUser: 50000
        fsGroup: 50000
      volumes:
        - name: dags
          emptyDir: {}
        - name: config
          configMap:
            name: airflow2-config
```
We see the issue a few times a day usually.

This is our k8s version deployed on Amazon EKS:
```
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.12-eks-a64ea69", GitCommit:"d4336843ba36120e9ed1491fddff5f2fec33eb77", GitTreeState:"clean", BuildDate:"2022-05-12T18:29:27Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

I have tried changing the config `hostname_callable = socket.gethostname` but this causes no improvement.

The pod name in this example is `hourlyreportingbidrequestssspdayhourlybidrequests.05d7664cfd9144469977dc191d35312d` which does not match the host name logged when the pod starts up, perhaps this is the cause of the mismatch? I'm not sure why the pod name would be different to the hostname that Airflow is aware of though.

**Reply by potiuk** [reply]
Hey @smartnewsdingli @cloventt . I have **just** merged a change https://github.com/apache/airflow/pull/24981 that has been inspired by this and similar issues people had in the past. 

I am not sure if the new "getfqdn" variant I added will solve the problems, but it has a good chance to do that.

And maybe It would be great - since you keep on having the problems recently you could try it out, before we release it? It should be rather easy - simply add this code somewhere in your local_settings or dags and configure `hostname_callable` to use it - and see if it helped?

```python
import socket
from functools import lru_cache

# patched version of socket.getfqdn() - see https://github.com/python/cpython/issues/49254
@lru_cache(maxsize=None)
def getfqdn(name=''):
    """Get fully qualified domain name from name.
    An empty argument is interpreted as meaning the local host.
    """
    name = name.strip()
    if not name or name == '0.0.0.0':
        name = socket.gethostname()
    try:
        addrs = socket.getaddrinfo(name, None, 0, socket.SOCK_DGRAM, 0, socket.AI_CANONNAME)
    except OSError:
        pass
    else:
        for addr in addrs:
            if addr[3]:
                name = addr[3]
                break
    return name
```

Would that be possible you try it and see if it helps?

**Reply by Swalloow** [reply]
```
The recorded hostname airflow-prod-worker-6656f79886-rgf5g 
does not match this instance's hostname airflow-prod-worker-6656f79886-cq46j
```
Is this message the same problem?
In my case, this happens more often with long-running worker pods.
We are using CeleryExecutor on EKS cluster.

**Reply by cloventt** [reply]
Kia ora @potiuk , thank you for contributing this change. I have patched our cluster (on version 2.3.3) to include the code and configured it to use the new method to determine the worker pod hostname.

Unfortunately this doesn’t seem to have resulted in any improvement in my case. I still see occasional errors that the hostname of the job runner does not match.

**Reply by SamanwayB** [reply]
Following this discussion since we are also facing the same.
We are having 25000 tasks per day, and getting this failure in 5-10 instances daily.

The old pod is getting evicted in Error status in Kubernetes and the new pod is getting completed by the scheduler with SIGTERM failure. 

Wondering if there is a possibility to change this to an issue.

We are using airflow 2.3.3 with no patches related to this.

**Reply by potiuk** [reply]
Try to upgrade to latest 2.4 maybe first ? And if it is still there providing similar logs and maybe more circumstances/correlation between what happens and the error, and monitoring (memory/cpu/ other k8s resources? ) the more  - the better.

**Reply by czammar02** [reply]
My team was facing the same problem, so temporary we increased the value of `AIRFLOW__SCHEDULER__JOB_HEARTBEAT_SEC` so our DAG's can end without being de terminated by the script `local_task_job.py `.

Another nasty solution I found was just to comment the line that rises the exception within the Dockerfile:
```
... 

# Temporary fix on hostname matching error
RUN cat /home/airflow/.local/lib/python3.7/site-packages/airflow/jobs/local_task_job.py | \
  sed 's/raise AirflowException("Hostname of job runner does not match")/#raise AirflowException("Hostname of job runner does not match")/g' \
  >> /home/airflow/.local/lib/python3.7/site-packages/airflow/jobs/local_task_job.py

```

Does anybody knows a better solution?

**Reply by rhames93** [reply]
I'm facing the same issue. I'm currently on 2.5.0 running 3 schedulers for about 200 dags (30 tasks each) for hourly schedule. We see 15-20 tasks fail every day due to the same error.

**Reply by potiuk** [reply]
Can you pleese open an issue and provide all the details of it - including details of logs from your k8S @rhames93? That would be immensely helpful if you can get as many details you can get from your k8S around the time when it happens.

Creating an issue for it might actually help in someone focusing on trying to diagnose it - the issue was converted to discussion because it had not enough details/likely reproducibility/enough input from the user. Seems you have it happening frequently enough  to be able to provide more details and open a more complete issue with beter chance to diagnose it.

**Comment by biancaberdugo** [comment]
Do we have updates on this issue?

**Reply by potiuk** [reply]
Generally speaking if ther are no updates here, then it means there are no updates. Nobody hides anything behind the scene in open-source project -if there  are no updates on issues / discussions, it means there are .... no updates. So asking if there are no updates does not change much.

But if your intention is "can I do something to make this issue addressed" - by all means you absolutely can @biancaberdugo . For example if you have similar issue in newer version of airlfow, you can provide more evidences, explain your case, show logs from newer version of Airflow. Or if you can reproduce similar case easily - you can open an issue with reproduction steps.

There are plenty of things you can do to help to resolve the issue, but asking "are there any updates?" is not one of those things.
